### PR TITLE
[MXNET-346] Hard Sigmoid Operator

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
Implementation of Hard Sigmoid function

### Essentials ###
- [x] The PR title starts with [MXNET-346]
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Hard Sigmoid on CPU and GPU
- [x] Unit test for Hard Sigmoid

## Comments ##
For bridging the operator gaps between MXNet and ONNX

### CODE REVIEW

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 1]
```diff
@@ -106,6 +106,35 @@ The storage type of ``sigmoid`` output is always dense
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_sigmoid,
                                                unary_bwd<mshadow_op::sigmoid_grad>);
+
+DMLC_REGISTER_PARAMETER(HardSigmoidParam);
+MXNET_OPERATOR_REGISTER_UNARY(hard_sigmoid)
+.describe(R"code(Computes hard sigmoid of x element-wise.
+
```
Comment (anirudhacharya) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: You could add an example here.

Comment (haojin2) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: It's unnecessary for this kind of function, please refer to "sigmoid" function above.

**Comment by haojin2** [comment]
@rahul003 @reminisce @anirudh2290 @eric-haibin-lin Could you please give a review for this?

**Comment by chinakook** [comment]
It’s widely used in LSTM in Keras. Could add this op into LSTM or GRU, etc..

**Comment by haojin2** [comment]
@chinakook Thanks for your advice! I'll definitely talk to related persons about this.

File: src/operator/tensor/elemwise_unary_op.h [hunk 2]
```diff
@@ -388,6 +388,94 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+      DType out_val = DType(alpha) * in_data[i] + DType(beta);
+      DType grad = (out_val > DType(0) && out_val < DType(1)) ?
+                   (out_grad[i] * DType(alpha)) : DType(0);
+      KERNEL_ASSIGN(in_grad[i], req, grad);
+    }
+};
+
+
+template<typename xpu>
+void HardSigmoidForward(const nnvm::NodeAttrs& attrs,
```
Comment (piiswrong) [inline src/operator/tensor/elemwise_unary_op.h]: why do you need to write forward and backward? You can use unary compute like https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/elemwise_unary_op_basic.cc#L87

Comment (piiswrong) [reply src/operator/tensor/elemwise_unary_op.h]: ok

Comment (haojin2) [reply src/operator/tensor/elemwise_unary_op.h]: So it's okay to leave this in this way?

File: src/operator/tensor/elemwise_unary_op.h [hunk 3]
```diff
@@ -388,6 +388,94 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+    .set_default(0.2)
+    .describe("Slope of hard sigmoid");
+    DMLC_DECLARE_FIELD(beta)
+    .set_default(0.5)
+    .describe("Bias of hard sigmoid.");
+  }
+};
+
+template<int req>
+struct hard_sigmoid_forward {
```
Comment (piiswrong) [inline src/operator/tensor/elemwise_unary_op.h]: please inherit tunable for fwd and bwd and add a entry here https://github.com/apache/incubator-mxnet/blob/daf821c09b3df581399a564a1b40b0a2e68750c9/src/operator/operator_tune.cc#L347

see https://github.com/apache/incubator-mxnet/blob/master/src/operator/mxnet_op.h#L596

Comment (piiswrong) [reply src/operator/tensor/elemwise_unary_op.h]: @cjolivier01  would know more about this. He is drafting a doc for autotune. I'll merge this for now. Please revisit after you go through that doc.

Comment (haojin2) [reply src/operator/tensor/elemwise_unary_op.h]: Sure, I actually talked with him yesterday after seeing your comments and those were what he told me, I think if we see more similar functions with this one then we can definitely get that done.

**Pull request merged**